### PR TITLE
Add support for BallTree, BinaryTree

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Those are used by `KNeighbors{Classifier,Regressor}` but were not caught so far because of the default argument.

Unfortunately, we need some special treatment here because the two cython classes, as well as `DistanceMetric`, which is used under the hood, return tuples from `__getstate__` and thus the normal `reduce_get_state` and `reduce_get_instance` methods fail.

LMK if you have a better name than `reduce_based_on_tuple_get_{state,instance}`.

Ready for review @adrinjalali

### CODE REVIEW

File: skops/io/_sklearn.py [hunk 1]
```diff
@@ -171,14 +176,52 @@ def reduce_get_instance(state, src, constructor):
...
     if cls not in ALLOWED_SGD_LOSSES:
         raise TypeError(f"Expected LossFunction, got {cls}")
-    return reduce_get_instance(state, src, cls)
+    return reduce_get_instance(state, src, constructor=cls)
+
+
+def reduce_based_on_tuple_get_state(obj, dst):
+    # For more details, take a look at reduce_get_state. This extra function is
+    # needed because __getstate__ of certain sklearn classes returns a tuple
+    # instead of a dict, which we thus need to treat differently.
```
Comment (adrinjalali) [inline skops/io/_sklearn.py]: why not change reduce so that it can handle a tuple/list?

File: skops/io/_sklearn.py [hunk 2]
```diff
@@ -13,6 +13,9 @@
     SquaredHinge,
     SquaredLoss,
 )
+from sklearn.metrics import DistanceMetric
+from sklearn.metrics._dist_metrics import METRIC_MAPPING
```
Comment (adrinjalali) [inline skops/io/_sklearn.py]: wouldn't maybe a user be able to modify this and hence allow things which shouldn't be allowed?

File: skops/io/_sklearn.py (hunk 1 above)
Comment (BenjaminBossan) [reply skops/io/_sklearn.py]: I was thinking about that but was afraid that the function, which is already quite complex, would become too confusing. If you think it's okay, I can work it in.

File: skops/io/_sklearn.py (hunk 2 above)
Comment (BenjaminBossan) [reply skops/io/_sklearn.py]: So would you suggest to hard-code the metrics instead?

Comment (adrinjalali) [reply skops/io/_sklearn.py]: yeah, like I did for losses.

File: skops/io/_sklearn.py (hunk 1 above)
Comment (adrinjalali) [reply skops/io/_sklearn.py]: No I think that function deserves to handle this complexity.

**Comment by BenjaminBossan** [comment]
@adrinjalali I pushed the change of hard-coding the distance metrics.

File: skops/io/_sklearn.py (hunk 1 above)
Comment (BenjaminBossan) [reply skops/io/_sklearn.py]: Is this a "no, don't merge the functions" or a "no, don't keep them separate"? :D

Comment (adrinjalali) [reply skops/io/_sklearn.py]: lolol, very fair. As in: I think the basic function should handle tuples, so please merge them together and have tuple handling in `reduce_*`.

**Comment by BenjaminBossan** [comment]
@adrinjalali I addressed your points, please review again.

File: skops/io/_sklearn.py [hunk 3]
```diff
@@ -144,10 +189,19 @@ def reduce_get_state(obj, dst):
...
+        content["args"] = {
+            "__class__": obj.__class__.__name__,
+            "__module__": get_module(type(obj)),
+            "__reduce__": {"state": _get_state(attrs, dst)},
+        }
+    else:
+        for key, value in attrs.items():
+            if isinstance(getattr(type(obj), key, None), property):
+                continue
+            content[key] = _get_state(value, dst)
```
Comment (adrinjalali) [inline skops/io/_sklearn.py]: can't we simply do:

```python
content["attrs"] = _get_state(attrs, dst)
```

and in the `reduce_get_instance`:

```python
args = _get_instance(state["content"]["attrs"], src)

if isinstance(args, tuple) and not hasattr(instance, "__setstate__"):
    raise

if hasattr(instance, "__setstate__"):
    instance.__setstate__(attrs)
else:
    instance.__dict__.update(attrs)
```

**Review by adrinjalali — COMMENTED** [review]
WDYT

**Comment by BenjaminBossan** [comment]
@adrinjalali Your suggestion wasn't quite enough, but it pointed me towards a huge simplification :+1: Please take a look again.

**Review by adrinjalali — APPROVED** [review]
Very nice!

**Pull request merged**